DataFormats/Detectors/CTP/src/Configuration.cxx

```diff
@@ -36,13 +40,186 @@ void CTPDetector::printStream(std::ostream& stream) const
 }
 void CTPCluster::printStream(std::ostream& stream) const
 {
-  stream << "CTP Cluster:" << name << " Inputs:" << detectorsMask << std::endl;
+  stream << "CTP Cluster:" << name << " mask:0b" << std::hex << detectorsMask << " "<< std::dec;
+  for(const auto &det : detectors){ stream << det.getName() << " ";}
+  stream << std::endl;
 }
 void CTPClass::printStream(std::ostream& stream) const
 {
-  stream << "CTP Class:" << name << " Hardware mask:" << classMask << std::endl;
+  stream << "CTP Class:" << name << " Hardware mask:" << classMask << " Descriptor:" << descriptor->name << " Cluster:"<< cluster->name << std::endl;
 }
 /// CTP configuration
+/// Assuming Run2 format + LTG
+int CTPConfiguration::loadConfiguration(std::string& ctpconfiguration)
+{
+  std::cout << "Loading CTP configuration." << std::endl;  
+  std::istringstream iss(ctpconfiguration);
+  int ret = 0;
+  int level=0;
+  std::string line;
+  while(std::getline(iss,line))
+  {
+    o2::utils::Str::trim(line);
+    if( (ret = processConfigurationLine(line,level)) !=0 )
+    {
+        return ret;
+    }
+  }
+  return ret;    
+}
+int CTPConfiguration::processConfigurationLine(std::string& line, int& level)
+{
+  if(line.size() == 0) return 0;
+  if(line.at(0) == '#') return 0;
+  size_t first;  
+  if((first=line.find("PARTITION:")) != std::string::npos)
+  {
+    mName = o2::utils::Str::trim_copy(line.erase(first,10));
+    return 0;
+  }
+  if((first=line.find("VERSION:")) != std::string::npos)
+  {
+    mVersion = o2::utils::Str::trim_copy(line.erase(first,8));
+    return 0;
+  }
+  if((first=line.find("INPUTS:")) != std::string::npos)
+  {
+    level = 1;
+    return 0;
+  }
+  if((first=line.find("DESCRIPTORS:")) != std::string::npos)
+  {
+    level = 3;
+    return 0;
+  }
+  if((first=line.find("CLUSTERS:")) != std::string::npos)
+  {
+      level = 4;
+      return 0;
+  }
+  if((first=line.find("CLASSES:")) != std::string::npos)
+  {
+    level = 7;
+    return 0;
+  }
+  /// Do parse levels
+  std::vector<std::string> tokens = o2::utils::Str::tokenize(line,' ');
+  size_t ntokens=tokens.size();
+  if(ntokens == 0) return 0;
+  switch (level)
+  {
+    case 1:
+    /// INPUTS: name det level indexCTP<0:45>
+    {
+        if(ntokens != 4)
+        {
+            std::cout << "Syntax error in INPUTS:" << line << std::endl;
+            return level;
+        }
+        CTPInput inp;
+        inp.name = tokens[0];
+        inp.detName = tokens[1];
+        inp.level = tokens[2];
+        o2::detectors::DetID det(inp.detName.c_str());
+        //std::cout << "id:" << det.getID() << " ndets:" << det.getNDetectors() << std::endl;
+        if(det.getID() >= det.getNDetectors())
+        {
+            std::cout << "INPUT line:" << line << " Detector does not exist: " << inp.detName << std::endl;
```

Comment:
better throw exception: `throw std::runtime_error(fmt::format("INPUT line {}: Detector {} does not exist", line, inp.detName);`